ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs

```diff
@@ -567,6 +571,46 @@ releaseResources rr keys releaser = do
 withRegistry :: (IOLike m, HasCallStack) => (ResourceRegistry m -> m a) -> m a
 withRegistry = bracket unsafeNewRegistry closeRegistry
 
+{-------------------------------------------------------------------------------
+  Temporary registry
+-------------------------------------------------------------------------------}
+
+-- | A temporary 'ResourceRegistry'
+--
+-- A registry to temporary allocate resources in before they make it to their
+-- final destination, i.e., the end of the scope of
+-- 'withTempResourceRegistry'. If an exception is thrown before the resources
+-- make it to their final destination, they should be cleaned up to prevent
+-- leaking them. If no exception is thrown and the resources safely make it to
+-- their final destination, the temporary registry does not have to release
+-- the resources.
+--
+-- For example, you're opening a file handle that you want to store in the
+-- state of the database. The file handle stored in the state of the database
+-- will be closed when the database closes. However, when an exception is
+-- thrown after opening the handle but /before/ it is stored in the database's
+-- state, the handle would leak. By using a 'TempResourceRegistry' to allocate
+-- the handle, we avoid the leak.
+newtype TempResourceRegistry m = UnsafeTempResourceRegistry (ResourceRegistry m)
+
+-- | A unidirectional pattern synonym to extract the underlying
+-- 'ResourceRegistry' without giving the ability to construct a
+-- 'TempResourceRegistry'.
+pattern TempResourceRegistry :: ResourceRegistry m -> TempResourceRegistry m
+pattern TempResourceRegistry registry <- UnsafeTempResourceRegistry registry
+{-# COMPLETE TempResourceRegistry #-}
+
+-- | Like 'withRegistry', but it only releases the resources in case of an
+-- exception. See 'TempResourceRegistry' for more details.
+withTempResourceRegistry
+  :: (IOLike m, HasCallStack)
+  => (TempResourceRegistry m -> m a)
+  -> m a
+withTempResourceRegistry =
+    bracketOnError
+      (UnsafeTempResourceRegistry <$> unsafeNewRegistry)
+      (\(TempResourceRegistry registry) -> closeRegistry registry)
+
```

Comment:
@kderme After thinking some more about the `withRegistryOnError` you introduced in #1681 (https://github.com/input-output-hk/ouroboros-network/commit/0d935923c3eba0736d22134cae9a0f21c1e8814f), @edsko and I realised it was actually a good idea.

In the VolatileDB, we no longer need to be AsyncException-resistant (which is something much more complex) thanks to #1709.